File: lib/fluxsource/hardwarefluxsource.cc
#include "lib/core/globals.h"
#include "lib/config/config.h"
#include "lib/config/flags.h"
#include "lib/data/fluxmap.h"
#include "lib/core/logger.h"
#include "lib/config/proto.h"
#include "lib/usb/usb.h"
#include "lib/fluxsource/fluxsource.h"
#include "lib/fluxsource/fluxsource.pb.h"

class HardwareFluxSource : public FluxSource
{
private:
    class HardwareFluxSourceIterator : public FluxSourceIterator
    {
    public:
        HardwareFluxSourceIterator(int track, int head):
            _track(track),
            _head(head)
        {
        }

        bool hasNext() const override
        {
            return true;
        }

        std::unique_ptr<const Fluxmap> next() override
        {
            const auto& drive = globalConfig()->drive();

            usbSetDrive(
                drive.drive(), drive.high_density(), drive.index_mode());
            usbSeek(_track);

            Bytes data = usbRead(_head,
                drive.sync_with_index(),
                drive.revolutions() * drive.rotational_period_ms() * 1e6,
                drive.hard_sector_threshold_ns());
            auto fluxmap = std::make_unique<Fluxmap>();
            fluxmap->appendBytes(data);
            return fluxmap;
        }

    private:
        int _track;
        int _head;
    };

public:
    HardwareFluxSource(const HardwareFluxSourceProto& conf): _config(conf) {}

    ~HardwareFluxSource() {}

public:
    std::unique_ptr<FluxSourceIterator> readFlux(int track, int head) override
    {
        return std::make_unique<HardwareFluxSourceIterator>(track, head);
    }

    void recalibrate() override
    {
        usbRecalibrate();
    }

    void seek(int track) override
    {
        usbSeek(track);
    }

    bool isHardware() override
    {
        return true;
    }

private:
    const HardwareFluxSourceProto& _config;
    bool _measured;
};

std::unique_ptr<FluxSource> FluxSource::createHardwareFluxSource(
    const HardwareFluxSourceProto& config)
{
    return std::unique_ptr<FluxSource>(new HardwareFluxSource(config));
}

File: java/com/cowlark/fluxengine/fluxsource/HardwareFluxSource.java
package com.cowlark.fluxengine.fluxsource;

import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.usb.UsbFactory;

/**
 * A flux source which reads from real hardware, ported from
 * lib/fluxsource/hardwarefluxsource.cc.
 */
public class HardwareFluxSource extends FluxSource
{
    private final ConfigProto config;
    private UsbFactory usbFactory;

    /* Package-private for testing. */
    HardwareFluxSource(ConfigProto config, UsbFactory usbFactory)
    {
        this.config = config;
        this.usbFactory = usbFactory;
    }

    @Override
    public FluxSourceIterator readFlux(FluxReadParameters parameters)
    {
        return new FluxSourceIterator()
        {
            @Override
            public boolean hasNext()
            {
                return true;
            }

            @Override
            public Fluxmap next()
            {
                Bytes[] box = new Bytes[1];
                usbFactory.perform(device -> box[0] = device.read(
                        parameters.cylinder(),
                        parameters.head(),
                        parameters.readTimeNs()));
                Fluxmap fluxmap = new Fluxmap();
                fluxmap.appendBytes(box[0]);
                return fluxmap;
            }
        };
    }

    @Override
    public boolean isHardware()
    {
        return true;
    }
}